mono/metadata/cominterop.c

```diff
@@ -1595,24 +1581,30 @@ typedef struct
 #define MONO_E_DISPID_UNKNOWN      (gint32)-1
 
 int
-ves_icall_System_Runtime_InteropServices_Marshal_AddRefInternal (gpointer pUnk)
+ves_icall_System_Runtime_InteropServices_Marshal_AddRefInternal (MonoIUnknown *pUnk, MonoError *error)
 {
 	g_assert (pUnk);
-	return (*(MonoIUnknown**)pUnk)->AddRef(pUnk);
+	return pUnk->vtable->AddRef(pUnk);
 }
 
 int
-ves_icall_System_Runtime_InteropServices_Marshal_QueryInterfaceInternal (gpointer pUnk, gconstpointer riid, gpointer* ppv)
+ves_icall_System_Runtime_InteropServices_Marshal_QueryInterfaceInternal (MonoIUnknown *pUnk, gconstpointer riid, gpointer* ppv, MonoError *error)
 {
 	g_assert (pUnk);
-	return (*(MonoIUnknown**)pUnk)->QueryInterface(pUnk, riid, ppv);
+	return pUnk->vtable->QueryInterface(pUnk, riid, ppv);
```

Comment:
bikeshedding: I'd love a static `mono_IUnknown_QueryInterface` that's called from here and from `cominterop_get_interface_checked` so that we can have fewer situations where icalls are used as utility functions inside the runtime.

Reply:
There were minor variations that made it not obvious how to unify internal calls and icall.
They were related to:
 - check null?
 - assert null? The assert is redundant though with the C# caller. But people are usually ok-or-better with simple safe fast redundant asserts, as long as the code isn't overrun with them.
 - use return value?
I thought I ended up with some unification but maybe more could be had.
I had to invent some names too, like mono_com_foo or mono_cominterop_foo or mono_IUknown_foo. IUnknown_Release and such are macros in the C versions of midl output. I believe they commit the sin of double evaluation (#define Foo_Bar(foo) ((foo)->vtable->bar(foo)) so I went with mono_IUnknown_foo.

Anyway I can look again.

Reply:
Ok I think I see your  point.
ves_AddRef and ves_QueryInterface have internal versions, for Release does not.
Probably based on what used/needed.

Reply:
and ves don't call ves, understood.

Reply:
and non-ves calling ves; ves only called from C#/icall-def.h